lib/Sema/CheckedCSubst.cpp

```diff
@@ -75,18 +76,254 @@ ExprResult Sema::ActOnTypeApplication(ExprResult TypeFunc, SourceLocation Loc,
+    // struct List _For_any(T) {
+    //   struct List<T> *next;
+    //   T *head;
+    // };
+    // While processing 'next', we can't instantiate 'List<T>' because we haven't processed the 'head' field yet.
+    // The solution is to mark the application as "delayed" and "complete it" after we've parsed all the fields.
+    Inst->setDelayedTypeApp(true);
+    ctx.addDelayedTypeApp(Inst);
+  }
+
+  return Inst;
+}
+
+void Sema::CompleteTypeAppFields(RecordDecl *Incomplete) {
+  assert(Incomplete->isInstantiated() && "Only instantiated record decls can be completed");
+  assert(Incomplete->field_empty() && "Can't complete record decl with non-empty fields");
+
+  auto Defn = Incomplete->baseDecl()->getDefinition();
+  assert(Defn && "The record definition should be populated at this point");
+  for (auto Field : Defn->fields()) {
+    QualType InstType = SubstituteTypeArgs(Field->getType(), Incomplete->typeArgs());
+    assert(!InstType.isNull() && "Subtitution of type args failed!");
+    // TODO(abeln): are TypeSouceInfo and InstType in sync?
+    FieldDecl* NewField = FieldDecl::Create(Field->getASTContext(), Incomplete, SourceLocation(), SourceLocation(),
+      Field->getIdentifier(), InstType, Field->getTypeSourceInfo(), Field->getBitWidth(), Field->isMutable(), Field->getInClassInitStyle());
+    Incomplete->addDecl(NewField);
+  }
+
+  Incomplete->setDelayedTypeApp(false);
+  Incomplete->setCompleteDefinition();
+}
+
+// Definitions for expanding cycles check
+namespace {
+  // A graph node is a triple '(BaseRecordDecl, TypeArgIndex, Expanding)' (here stored as two nested 'std::pairs' so we can keep them in a 'DenseSet').
+  // The semantics of a triple are as follows: a triple is in the set if, starting from one of the type arguments of 'Base', 
+  // it's possible to arrive at the type argument with index 'TypeArgIndex' which is defined in 'BaseRecordDecl'.
+  // 'Expanding' indicates whether any of the edges taken to arrive to (BaseRecordDecl, TypeArgIndex) is expanding ('Expanding = 1')
+  // or if they're all non-expanding ('Expanding = 0'). 
+  using Node = std::pair<std::pair<RecordDecl *, int>, char>;
+  const char NON_EXPANDING = 0;
+  const char EXPANDING = 1;
+
+  // Retrieve the underlying type variable from a typedef that appears as the param to a generic record.
+  const TypeVariableType *GetTypeVar(TypedefDecl *TDef) {
+    const TypeVariableType *TypeVar = llvm::dyn_cast<TypeVariableType>(TDef->getUnderlyingType());
+    assert(TypeVar && "Expected a type variable as the parameter of a generic record");
+    return TypeVar;
+  }
+
+  // A 'TypeVisitor' that determines whether a type references a given type variable.
+  // e.g. ContainsTypeVar('T').Visit('List<T>') -> true
+  //      ContainsTypeVar('T').Visit('List<int>') -> false
+  class ContainsTypeVarVisitor : public TypeVisitor<ContainsTypeVarVisitor, bool> {
```

Comment:
> I think you should change this pattern to not rely on the return type. I'm not sure that the return value will be propagated for all kinds of types. The typical pattern that is used is to have a field that is set to true if the type variable is encountered. Then after invoking the traversel, you check that field.

I tried this, but since we re-use the visitor multiple times in a loop, we'd also have to add a method to "reset" the results of the previous run. I think having the visitor be stateless makes for a nicer api.

> 
> I think this wil let you delete some methods. I believe it will also handle some missing cases, like function types with a recursive type application embedded in them.

Unfortunately, TypeVisitor doesn't structurally recurse on the components of a type by default. The default implementation just returns false. We have to handle _every_ case we care about by hand.

Reply:
I didn't realize that TypeVisitor doesn't recursively traverse types that are components of otehr types.  In that case, I think you should switch to RecursiveASTVisitor, which does.   The current code pattern is dangerous because it is easy to miss a case.   If that happens, we won't notice the use of the type variable.  

For the re-use, we should just create a new object for each iteration of the loop.  The object doesn't need to be heap-allocated, it can be stack-allocated.

Reply:
Switched to RecursiveASTVisitor. I was able to remove a few of the cases, but not all of them (e.g. RecursiveASTVisitor doesn't handle TypeDefType and RecordType).

I also had to move the expanding cycles visitors to their own file, because the obj file for CheckedCSubst had grown too large and couldn't be generated. I was getting
```
fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj split file
``

Splitting the file seems to solve the problem.

Reply:
As per David's suggestion, I moved the expanding cycles code back to CheckedCSubst, and modified the Cmake file.